### ISSUE TITLE
Serializer validation. Issue with parsing dictionary if array is a value

### ISSUE THREAD

**Issue by akolpakov** [tomchristie/django-rest-framework#3598]
[At this line](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/utils/html.py#L65) function ``parse_html_dict`` allow to parse dictionaries with primitive values like:
```
    {
        'profile.username': 'example',
        'profile.email': 'example@example.com',
    }
        -->
    {
        'profile': {
            'username': 'example',
            'email': 'example@example.com'
        }
    }
```

But if value is an array parser does not work properly:
```
    {
        'profile.username': 'example',
        'profile.email': 'example@example.com',
        'profile.group': ['1', '2'],
    }
```

Problem that QueryDict on the [line](https://github.com/tomchristie/django-rest-framework/blob/master/rest_framework/utils/html.py#L83):
```python
for field, value in dictionary.items():
```

``dictionary.items()`` returns last item from list ['1', '2'] as a ``value``.

I have the following Serializer:

```python
class ProfileSerializer(serializers.ModelSerializer):
    class Meta:
        model = Profile

class UserSerializer(serializers.ModelSerializer):
    profile = ProfileSerializer(required=False)
```

and ``group`` is a ``ManyToMany`` field in ``Profile`` model.

**Comment by dursk** [comment]
Having trouble reproducing what you're describing. The `parse_html_dict` seems to be working as expected.
```python
>>> from rest_framework.utils.html import parse_html_dict
>>> input_ = {'profile.username': 'example', 'profile.email': 'foo@bar.com', 'profile.group': ['1', '2']}
>>> output = parse_html_dict(input_, prefix='profile')
>>> output
<MultiValueDict: {'username': ['example'], 'group': [['1', '2']], 'email': ['foo@bar.com']}>
>>> output['group']
['1', '2']
```

**Comment by tomchristie** [comment]
Closing pending a reproducible example - thanks @dursk!
@akolpakov we can reassess this if you can nail this down further.

**Comment by akolpakov** [comment]
Guys, I've made a test case for this issue (pull request #3710).
I don't know how to fix it in a good way. Please, look at it.

**Comment by lrnt** [comment]
I fixed this issue in pull request #3820, it also includes the necessary tests. This fix also works for the test case added by pull request #3710 from @akolpakov.

@dursk: `parse_html_dict` expects a [`QueryDict`](https://docs.djangoproject.com/en/stable/ref/request-response/#querydict-objects) or `MultipleValueDict` as parameter, which behaves a little bit differently than a regular dictionary. Both of these datatypes can have multiple values for one key. When you try to access that key it will return the last value assigned to the key. This is problematic when dealing with lists.

A more appropriate quick test would be:
```python
>>> from rest_framework.utils.html import parse_html_dict
>>> from django.utils.datastructures import MultiValueDict
>>> input = MultiValueDict({'profile.username': 'example', 'profile.email': 'foo@bar.com', 'profile.group': ['1', '2']})
>>> output = parse_html_dict(input, prefix='profile')
>>> output
<MultiValueDict: {'email': ['m'], 'username': ['e'], 'group': ['2']}>
>>> output['group']
'2'
```

Please consider reopening this issue.

**Comment by tomchristie** [comment]
Have milestoned the test case and fix accordingly to bump them up our priority list.
